client-sdk/ts-web/core/src/hdkey.ts

```diff
@@ -0,0 +1,89 @@
+import {sha512} from 'js-sha512';
+import {SignKeyPair, sign} from 'tweetnacl';
+import {generateMnemonic, mnemonicToSeed} from 'bip39';
+
+const ED25519_CURVE = 'ed25519 seed';
+const HARDENED_OFFSET = 0x80000000;
+const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
+
+export class HDKey {
+    public readonly keypair: SignKeyPair;
+
+    constructor(public readonly privateKey: Uint8Array, public readonly chainCode: Uint8Array) {
+        this.keypair = sign.keyPair.fromSeed(privateKey);
+    }
+
+    /**
+     * Returns an instance of HDKey that can then be used for derivation
+     * @param seed Master seed for key derivation
+     * @returns Instance of HDKey
+     */
+    public static fromSeed(seed: Uint8Array): HDKey {
+        return HDKey.makeHDKey(ED25519_CURVE, seed);
+    }
+
+    /**
+     * Returns the root node for the given mnemonic
+     * @param mnemonic Bip39 Mnemonic
+     * @param passphrase Optional passphrase
+     * @returns Root node
+     */
+    public static async fromMnemonic(mnemonic: string, passphrase?: string): Promise<HDKey> {
+        const seed = await mnemonicToSeed(mnemonic, passphrase);
+        return HDKey.fromSeed(seed);
+    }
+
+    /**
+     * Generates a mnemonic
+     * @param strength Length in bits of the generated mnemonic
+     * @returns Generated BIP39 Mnemonic
+     */
+    public static generateMnemonic(strength = 256): string {
+        return generateMnemonic(strength);
+    }
+
+    /**
+     * Returns the HDKey for the given derivation path
+     * @param path Derivation path, starting with m/
+     * @returns Instance of HDKey
+     */
+    public derivePath(path: string): HDKey {
+        if (!pathRegex.test(path)) {
+            throw new Error(
+                "Invalid derivation path. Valid paths must use a format similar to : m/44'/474'/0' and all indexes must be hardened",
+            );
+        }
+
+        const segments = path
+            .split('/')
+            .slice(1)
+            .map((val: string): string => val.replace("'", ''))
+            .map((el) => parseInt(el, 10));
+
+        return segments.reduce((parent, segment) => parent.derive(segment + HARDENED_OFFSET), this);
+    }
```

Comment:
I think it would make sense to have a high-level function analogous to (taken from https://github.com/oasisprotocol/oasis-core/pull/3918/):
```go
// Generate signer for the given mnemonic, passphrase and account according to
// ADR 0008.
func GetAccountSigner(
	mnemonic string,
	passphrase string,
	number uint32,
) (signature.Signer, BIP32Path, error) {
...
}
```
In this case, it wouldn't return a signer but the `SignKeyPair` object.

That way, ts-web users don't need to worry about the low-level BIP-0032 paths but can just access Oasis accounts by their numbers (for a given wallet).